tests/algorithms/test_adam.py

```diff
@@ -0,0 +1,156 @@
+"""Tests for Adam."""
+
+# Copyright 2023 Physikalisch-Technische Bundesanstalt
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#       http://www.apache.org/licenses/LICENSE-2.0
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import pytest
+import torch
+import torch.nn.functional as F
+
+from mrpro.algorithms import adam
+from mrpro.data import SpatialDimension
+from mrpro.operators import ConstraintsOp
+from mrpro.operators.functionals import L2_data_discrepancy
+from tests import RandomGenerator
+from tests.operators._OptimizationTestFunctions import Rosenbrock
+from tests.phantoms._SaturationRecoveryEllipsePhantom import SaturationRecoveryEllipsePhantom2D
+
+
+@pytest.mark.parametrize(
+    'a, b',
+    [(1.0, 100.0), (2.0, 50.0), (5.0, 10.0), (10.0, 5.0)],
+)
+def test_adam_rosenbrock(a, b):
+    """Test adam functionality."""
+
+    with pytest.raises(ImportWarning):  # ToDo: remove this when fixed
+        # hyperparams or adam
+        lr = 1e-2
+        max_iter = 1000
+
+        random_generator = RandomGenerator(seed=0)
+
+        # generate two-dimensional test data
+        x1 = random_generator.float32_tensor(size=(1,))
+        x2 = random_generator.float32_tensor(size=(1,))
+
+        # enable gradient calculation
+        x1.requires_grad = True
+        x2.requires_grad = True
+
+        params = [x1, x2]
+
+        # define Rosenbrock function
+        f = Rosenbrock(a, b)
+
+        # call adam
+        params = adam(
+            f,
+            params,
+            max_iter=max_iter,
+            lr=lr,
+        )
+
+        # minimizer of Rosenbrock function
+        sol = torch.tensor([a, a**2])
+
+        # obtained solution
+        x = torch.tensor(params)
+
+        # check if they are close
+        tol = 1e-1
+        mse = F.mse_loss(sol, x)
+        assert mse < tol
+
+
+@pytest.mark.parametrize(
+    'ny, nx, bounds_flag',
+    [
+        (128, 128, 0),
```

Comment:
Does this parameterization make any sense?

Why should adam fail with a different number of pixels?
this only increases runtime of the tests